packages/format-library/src/link/inline.js

```diff
@@ -30,12 +38,71 @@ import { useDispatch, useSelect } from '@wordpress/data';
 import { createLinkFormat, isValidHref, getFormatBoundary } from './utils';
 import { link as settings } from './index';
 
+const TogglableSettingComponent = ( { setting, value, onChange } ) => {
+	const hasValue = value ? value?.cssClasses?.length > 0 : false;
+	const [ inputVisible, setInputVisible ] = useState( hasValue );
+
+	const handleSettingChange = ( newValue ) => {
+		onChange( {
+			...value,
+			[ setting.id ]: newValue,
+		} );
+	};
+
+	const handleCheckboxChange = () => {
+		if ( inputVisible ) {
+			if ( hasValue ) {
+				// Reset the value.
+				handleSettingChange( '' );
+			}
+			setInputVisible( false );
+		} else {
+			setInputVisible( true );
+		}
+	};
+
+	return (
+		<div className="block-editor-link-control__toggleable-setting">
+			<CheckboxControl
+				__nextHasNoMarginBottom
+				label={ setting.title }
+				onChange={ handleCheckboxChange }
+				checked={ inputVisible || hasValue }
+				help={ setting?.help }
+			/>
+			{ inputVisible && (
+				<InputControl
+					label={ setting.title }
+					value={ value?.cssClasses }
+					onChange={ handleSettingChange }
+					help={ __( 'Separate multiple classes with spaces.' ) }
+					__unstableInputWidth="100%"
+					__next40pxDefaultSize
+				/>
```

Comment:
I'm thinking we could do with some validation and sanitization at this point. There might be some prior art for this.